Calling `onos config snapshot devicesim-1` causes onos-config to crash

**Describe the bug**
On a newly installed system on `onit` with a few network changes and no snapshots, when I call
`onos config snapshot devicesim-1`
it causes onos-config to crash


**To Reproduce**
1. Create a cluster with `onit`
```bash
onit create cluster --set onos-cli.enabled=true
```
2. create a simulator `devicesim-1`:
```bash
onit -c onos add simulator -n devicesim-1
```
3. Run the snapshot command to retrieve snapshots
`onos config snapshot devicesim-1`

4. Observe that this command crashes onow config.
a) the reply is 
```bash
rpc error: code = Unavailable desc = transport is closing
```
b) kubectl shows that onos has restarted (see the restarts column):
```bash
Every 2.0s: kubectl get pods -n onos                                                                                                                       Z420: Fri Jan 10 10:02:09 2020

NAME                                 READY   STATUS    RESTARTS   AGE
atomix-controller-756f86cc9c-jjhhl   1/1     Running   0          15m
database-1-0                         1/1     Running   0          14m
devicesim-1                          1/1     Running   0          12m
devicesim-2                          1/1     Running   0          11m
onos-cli-8fddbd77f-c427r             1/1     Running   0          14m
onos-config-7f77c68c88-m7msc         1/1     Running   1          14m
onos-topo-58dbddcb64-rw8xc           1/1     Running   0          14m
```
c) onos-config stack trace below is produced

**Expected behavior**
I expected it to be able to handle the case where there are no snapshots on the system

**Logs, text or screenshots**
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x10dd0d6]

goroutine 574 [running]:
github.com/onosproject/onos-config/api/types/snapshot/device.(*Snapshot).MarshalToSizedBuffer(0x0, 0x1e8a588, 0x0, 0x0, 0x87c803, 0x1237ea0, 0x1328a60)
	/go/src/github.com/onosproject/onos-config/api/types/snapshot/device/types.pb.go:522 +0x26
github.com/onosproject/onos-config/api/types/snapshot/device.(*Snapshot).Marshal(0x0, 0x1328a60, 0x0, 0x7f22a3ea28b8, 0x0, 0xc0020a6001)
	/go/src/github.com/onosproject/onos-config/api/types/snapshot/device/types.pb.go:505 +0x83
google.golang.org/grpc/encoding/proto.codec.Marshal(0x1328a60, 0x0, 0xc000402540, 0x14974c0, 0x87b3cc, 0xc000204000, 0x1248ba0)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/encoding/proto/proto.go:70 +0x194
google.golang.org/grpc.encode(0x7f22a8dcd3b0, 0x1e8a588, 0x1328a60, 0x0, 0x1e8a588, 0xc003d5b790, 0x11d2980, 0xc003d5b790, 0x11dc340)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/rpc_util.go:543 +0x52
google.golang.org/grpc.(*Server).sendResponse(0xc003768c00, 0x14a1a80, 0xc000afbe00, 0xc003c7a000, 0x1328a60, 0x0, 0x0, 0x0, 0xc003eb9388, 0x0, ...)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:845 +0x8b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc003768c00, 0x14a1a80, 0xc000afbe00, 0xc003c7a000, 0xc00371ad50, 0x1e1cc58, 0x0, 0x0, 0x0)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:1042 +0x557
google.golang.org/grpc.(*Server).handleStream(0xc003768c00, 0x14a1a80, 0xc000afbe00, 0xc003c7a000, 0x0)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:1287 +0xda6
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0037e2fe0, 0xc003768c00, 0x14a1a80, 0xc000afbe00, 0xc003c7a000)
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:722 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:720 +0xa1
```

**Additional context**
Add any other context about the problem here.

--- a/pkg/northbound/admin/admin.go
+++ b/pkg/northbound/admin/admin.go
@@ -31,6 +31,8 @@ import (
 	"github.com/onosproject/onos-config/pkg/utils/logging"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 	"io"
 	"os"
 	"strings"
@@ -201,7 +203,19 @@ func (s Server) RollbackNetworkChange(ctx context.Context, req *admin.RollbackRe
 
 // GetSnapshot gets a snapshot for a specific device
 func (s Server) GetSnapshot(ctx context.Context, request *admin.GetSnapshotRequest) (*devicesnapshot.Snapshot, error) {
-	return manager.GetManager().DeviceSnapshotStore.Load(devicetype.NewVersionedID(request.DeviceID, request.DeviceVersion))
+	if len(request.DeviceID) == 0 || len(request.DeviceVersion) == 0 {
+		return nil, status.Error(codes.InvalidArgument,
+			fmt.Sprintf("An ID and Version must be given. Got ID: %s, Version: %s", request.DeviceID, request.DeviceVersion))
+	}
+
+	snapshot, err := manager.GetManager().DeviceSnapshotStore.Load(devicetype.NewVersionedID(request.DeviceID, request.DeviceVersion))
+	if err != nil {
+		return nil, err
+	}
+	if snapshot == nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("No snapshot found for %s:%s", request.DeviceID, request.DeviceVersion))
+	}
+	return snapshot, err
 }
 
 // ListSnapshots lists snapshots for all devices

--- a/pkg/northbound/admin/admin_test.go
+++ b/pkg/northbound/admin/admin_test.go
@@ -17,8 +17,12 @@ package admin
 import (
 	"context"
 	"errors"
+	"fmt"
 	"github.com/golang/mock/gomock"
 	"github.com/onosproject/onos-config/api/admin"
+	device2 "github.com/onosproject/onos-config/api/types/change/device"
+	"github.com/onosproject/onos-config/api/types/device"
+	devicesnapshot "github.com/onosproject/onos-config/api/types/snapshot/device"
 	"github.com/onosproject/onos-config/pkg/manager"
 	mockstore "github.com/onosproject/onos-config/pkg/test/mocks/store"
 	"github.com/onosproject/onos-config/pkg/test/mocks/store/cache"
@@ -101,4 +105,65 @@ func Test_RollbackNetworkChange_NoChange(t *testing.T) {
 	mockNwChStore.EXPECT().Get(gomock.Any()).Return(nil, errors.New("change is not specified"))
 	_, err := client.RollbackNetworkChange(context.Background(), &admin.RollbackRequest{Name: ""})
 	assert.ErrorContains(t, err, "is not")
+}
+
+func Test_GetSnapshot(t *testing.T) {
+	mgrTest, conn, client, server := setUpServer(t)
+	defer server.Stop()
+	defer conn.Close()
+
+	const dev1ID = "presentdevice"
+	const dev1Ver = "2.0.0"
+	snapshotID := devicesnapshot.ID(fmt.Sprintf("Snapshot:%s:%s", dev1ID, dev1Ver))
+
+	mockDevSnapshotStore, ok := mgrTest.DeviceSnapshotStore.(*mockstore.MockDeviceSnapshotStore)
+	assert.Assert(t, ok, "casting mock store")
+
+	mockDevSnapshotStore.EXPECT().Load(gomock.Any()).DoAndReturn(
+		func(devId device.VersionedID) (*devicesnapshot.Snapshot, error) {
+			if devId.GetID() == dev1ID && devId.GetVersion() == dev1Ver {
+				return &devicesnapshot.Snapshot{
+					ID:            snapshotID,
+					DeviceID:      dev1ID,
+					DeviceVersion: dev1Ver,
+					DeviceType:    "TestDevice",
+					SnapshotID:    "test-snapshot",
+					ChangeIndex:   0,
+					Values: []*device2.PathValue{
+						{Path: "/a/b/c", Value: device2.NewTypedValueInt64(12345)},
+						{Path: "/a/b/d", Value: device2.NewTypedValueInt64(12346)},
+					},
+				}, nil
+			}
+			return nil, nil
+		}).AnyTimes()
+
+	snapshot, err := client.GetSnapshot(context.Background(), &admin.GetSnapshotRequest{
+		DeviceID:      device.ID("testdevice"),
+		DeviceVersion: device.Version(""),
+	})
+	assert.ErrorContains(t, err, "An ID and Version must be given")
+	assert.Assert(t, snapshot == nil, "Expecting snapshot to be nil")
+
+	// Try again with a version
+	snapshot, err = client.GetSnapshot(context.Background(), &admin.GetSnapshotRequest{
+		DeviceID:      device.ID("absentdevice"),
+		DeviceVersion: device.Version("1.0.0"),
+	})
+	assert.ErrorContains(t, err, "No snapshot found")
+	assert.Assert(t, snapshot == nil, "Expecting snapshot to be nil")
+
+	// Try again with a positive response
+	snapshot, err = client.GetSnapshot(context.Background(), &admin.GetSnapshotRequest{
+		DeviceID:      dev1ID,
+		DeviceVersion: dev1Ver,
+	})
+	assert.NilError(t, err, "Not expecting error on GetSnapshot")
+
+	assert.Equal(t, string(snapshotID), string(snapshot.ID))
+	assert.Equal(t, dev1ID, string(snapshot.DeviceID))
+	assert.Equal(t, dev1Ver, string(snapshot.DeviceVersion))
+	assert.Equal(t, "TestDevice", string(snapshot.DeviceType))
+	assert.Equal(t, "test-snapshot", string(snapshot.SnapshotID))
+	assert.Equal(t, 2, len(snapshot.GetValues()))
 }

--- a/test/gnmi/snapshot.go
+++ b/test/gnmi/snapshot.go
@@ -1,0 +1,63 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gnmi
+
+import (
+	"context"
+	"github.com/onosproject/onos-config/api/admin"
+	"github.com/onosproject/onos-config/api/types/device"
+	"github.com/onosproject/onos-test/pkg/onit/env"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+// TestSnapshotErrors tests the Snapshot RPCs on the Admin gRPC interface
+func (s *TestSuite) TestSnapshotErrors(t *testing.T) {
+	// Data to run the test cases
+	testCases := []struct {
+		description   string
+		id            string
+		version       string
+		expectedError string
+	}{
+		{description: "No version", id: "devicesim-1", version: "", expectedError: "An ID and Version must be given"},
+		{description: "No snapshot", id: "devicesim-1", version: "1.0.0", expectedError: "No snapshot found"},
+	}
+
+	adminClient, err := env.Config().NewAdminServiceClient()
+	assert.NoError(t, err)
+
+	// Run the test cases
+	for _, testCase := range testCases {
+		thisTestCase := testCase
+		t.Run(thisTestCase.description,
+			func(t *testing.T) {
+				description := device.ID(thisTestCase.description)
+				id := device.ID(thisTestCase.id)
+				version := device.Version(thisTestCase.version)
+				expectedError := thisTestCase.expectedError
+
+				t.Logf("testing %q", description)
+
+				snapshot, err := adminClient.GetSnapshot(context.Background(), &admin.GetSnapshotRequest{
+					DeviceID:      id,
+					DeviceVersion: version,
+				})
+				assert.Nil(t, snapshot, "Expected snapshot to be nil")
+				assert.Error(t, err, "Expected error response", expectedError)
+			})
+	}
+
+}

